### PR TITLE
Project: Chess

### PR DESCRIPTION
**Pull request by Kaktug**
# Description

This PR contains my solution to The Odin Project's "Chess Game" project.  The assignment can be seen here:
https://www.theodinproject.com/lessons/ruby-ruby-final-project

> ## Assignment
> **1. Build a command line Chess game where two players can play against each other.**
> 2. The game should be properly constrained – it should prevent players from making illegal moves and declare check or check mate in the correct situations.
> 3. Make it so you can save the board at any time (remember how to serialize?)
> 4. Write tests for the important parts. You don’t need to TDD it (unless you want to), but be sure to use RSpec tests for anything that you find yourself typing into the command line repeatedly.
> 5. Do your best to keep your classes modular and clean and your methods doing only one thing each. This is the largest program that you’ve written, so you’ll definitely start to see the benefits of good organization (and testing) when you start running into bugs.
> 6. Unfamiliar with Chess? Check out some of the additional resources to help you get your bearings.
> 7. Have fun! Check out the [unicode characters](http://en.wikipedia.org/wiki/Chess_symbols_in_Unicode) for a little spice for your gameboard.
> 8. (Optional extension) Build a very simple AI computer player (perhaps who does a random legal move)

# Notes
This was a slog. 🫨  Some of the advanced rules (en-passant, castling) were really hard to shoehorn in, and I wish I'd focused on those earlier.  I feel OK about the piece and board classes, and the game seems to play alright, at least.

If you run the test suite you may want to comment out calls to the Display class' `clear()` and `welcome()` methods.  They mess up the terminal output a bunch in VSCode (scrolling back through the buffer works fine in iTerm2 though).

# To Play
1. Clone this repo and `cd` into the `project-chess` dir.
2. Run `ruby main.rb`

### CODE REVIEW

**Review by zacharydanger — COMMENTED** [review]
Very neat stuff!

Couple of UX things:

CleanShot 2023-05-09 at 10 03 35@2x

1. How does it want me to input moves?
2. You might remind the player which color they are. Maybe like `Zach (:king unicode char:)`?

File: full-stack-ruby-on-rails/project-chess/lib/pieces/bishop.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+# A Bishop piece in a chess game
+class Bishop < Piece
+  def possible_moves(board)
+    sliding_diagonal_moves(position, board)
+  end
+
+  def trapped?(board)
+    possible_moves(board).empty?
+  end
```
[the remark this answers is not in the document]
Comment (Kaktug) [reply full-stack-ruby-on-rails/project-chess/lib/pieces/bishop.rb:13]: That's a good idea.  The problem I was having is that the `Piece` class doesn't implement the `possible_moves` method -- that's something each (extended) piece implements itself.  Would usual practice be to stub `possible_moves` inside Piece?  Something like...:
```
def possible_moves(board)
  []
end
```

**Comment by Kaktug** [comment]
>     1. How does it want me to input moves?

Good question, I added some more detail to the movement prompt (specifically, providing an example of the expected notation and replaced the word 'Move' with 'Type').

TL;DR It expects chess notation, so for example: "a2 a4".  That command would select the piece in `a2` and try to move it to `a4` (if the move is valid and if there is a piece in a2).  Good feedbac
 
>     2. You might remind the player which color they are. Maybe like `Zach (:king unicode char:)`?

I like it.  Dunzo.

File: full-stack-ruby-on-rails/project-chess/lib/pieces/bishop.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (zacharydanger) [reply full-stack-ruby-on-rails/project-chess/lib/pieces/bishop.rb:13]: Yup! Another option is to have it raise an error so that it's _required_ to be implemented in the subclass.

File: full-stack-ruby-on-rails/project-chess/lib/board.rb [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  def special_pieces
+    default_special_pieces.each_with_object({}) do |piece, hash|
+      hash[piece.to_s.downcase] = piece
+    end
+  end
+
+  private
+
+  def default_special_pieces
+    [Rook, Knight, Bishop, Queen, King, Bishop, Knight, Rook]
```
Comment (chmurph2) [inline full-stack-ruby-on-rails/project-chess/lib/board.rb:91]: Interesting use of Class constants here.

File: full-stack-ruby-on-rails/project-chess/lib/chess.rb [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+  end
+
+  def array_to_chess_notation(array_notation)
+    row, column = array_notation
+    row = board.size - row
+    "#{indexed_alphabet.key(column)}#{row}"
+  end
+
+  def print_error(piece, destination)
+    if !own_piece?(piece)
```
Comment (chmurph2) [inline full-stack-ruby-on-rails/project-chess/lib/chess.rb:162]: For readability, I sometimes will create a helper method called `not_own_piece?` but that may just be a me-thing not liking negation with `!`.

**Review by chmurph2 — APPROVED** [review]
Good grief, what a big project they made you do! It's a lot, but I did my best to follow it. I really like how your methods are well named and generally small, making it all relatively easy to read. Some of the logic here I just... moved on from, I admit (not your fault). 

I'm not surprised most of the team didn't review such a big PR, and chose to Homer it back into the bush. Not much else you can do for this kind of project assignment. You already know we encourage smaller PRs IRL so nothing to worry about there.

File: full-stack-ruby-on-rails/project-chess/lib/chess.rb (hunk 3 above)
Comment (Kaktug) [reply full-stack-ruby-on-rails/project-chess/lib/chess.rb]: Ahh, thank you for pointing that out.  I've seen that in our code before and I like those convenience methods too...the `!` is easy to overlook and a little awkward when used like: `!board.check?`.
